src/modules/vault/components/buy-sell/BuyOrSellForm/SourceCurrency.tsx

```diff
@@ -0,0 +1,256 @@
+import {
+  Button,
+  ButtonGroup,
+  Flex,
+  InputGroup,
+  InputRightAddon,
+  Stack,
+  Text,
+  useDisclosure,
+} from '@chakra-ui/react';
+import { bn } from 'fuels';
+import { useMemo, useRef } from 'react';
+import { Controller, useFormContext } from 'react-hook-form';
+
+import { CurrencyCode, CurrencyField } from '@/components';
+import { SelectedCurrency } from '@/modules/core/components';
+import { ICreateWidgetPayload } from '@/modules/core/models/meld';
+import {
+  useListCryptoCurrencies,
+  useListFiatCurrencies,
+} from '@/modules/vault/hooks';
+import { parseToNumber, removeRightZeroes } from '@/modules/vault/utils';
+import { useVaultInfosContext } from '@/modules/vault/VaultInfosProvider';
+import { FUEL_ETH_ID } from '@/utils/constants';
+import mergeRefs from '@/utils/merge-refs';
+
+import { CardRoot } from '../CardRoot';
+import { CurrencyOptionsModal } from '../CurrencyOptionsModal';
+import { InputMirror } from '../InputMirror';
+
+export const SourceCurrency = ({
+  maxAmount,
+  minAmount,
+  isOnRamp,
+}: {
+  minAmount?: number;
+  maxAmount?: number;
+  isOnRamp: boolean;
+}) => {
+  const inputRef = useRef<HTMLInputElement>(null);
+  const { fiatCurrencies, isLoading: isLoadingCurrencies } =
+    useListFiatCurrencies();
+  const { assets } = useVaultInfosContext();
+  const { cryptoCurrencies, isLoading: isLoadingCryptoCurrencies } =
+    useListCryptoCurrencies();
+
+  const {
+    control,
+    watch,
+    setValue,
+    setFocus,
+    formState: { errors },
+  } = useFormContext<ICreateWidgetPayload>();
+  const currencyModal = useDisclosure();
+
+  const currentCurrencyCode = watch('sourceCurrencyCode');
+  const fuelEthAsset = useMemo(
+    () => assets.assets?.find((asset) => asset.assetId === FUEL_ETH_ID),
+    [assets],
+  );
+
+  const handleCurrencyChange = (currencyCode: string) => {
+    setValue('sourceCurrencyCode', currencyCode);
+    currencyModal.onClose();
+  };
+  const currencyOptions = isOnRamp ? fiatCurrencies : cryptoCurrencies;
+
+  const currentCurrency = useMemo(
+    () =>
+      currencyOptions.find(
+        (currency) => currency.currencyCode === currentCurrencyCode,
+      ),
+    [currencyOptions, currentCurrencyCode],
+  );
+
+  const isLoading = isLoadingCryptoCurrencies || isLoadingCurrencies;
+  const currencies = useMemo(
+    () =>
+      currencyOptions.map((currency) => ({
+        label: currency.name,
+        value: currency.currencyCode,
+        imageUrl: currency.symbolImageUrl,
+      })),
+    [currencyOptions],
+  );
+
+  const ethAmount = useMemo(
+    () => bn(fuelEthAsset?.amount ?? 0).formatUnits(9),
+    [fuelEthAsset],
+  );
+
+  const handleSetCurrencyAmount = (percentage: number) => {
+    const amount = parseToNumber(ethAmount) * (percentage / 100);
+    setValue('sourceAmount', amount.toString());
+    setFocus('sourceAmount');
+  };
```

Comment:
```suggestion
  const ethAmount = useMemo(
    () => bn(fuelEthAsset?.amount ?? 0),
    [fuelEthAsset],
  );

  const handleSetCurrencyAmount = (percentage: number) => {
    const amount = ethAmount.mul(percentage).div(100).add(ethAmount);
    setValue('sourceAmount', amount.format({ precision: 9 }));
    setFocus('sourceAmount');
  };
```

always use `bn` to work with BigNumbers

Reply:
Done!